README.md
# AWS Proxies

This allows to create AWS instances with multiple public ips to create proxies servers.

This is still under heavy development. Use at own risk.

## Installation

Using Pip:

```
$ pip install -e git+git@github.com:davidlonjon/aws proxies.git@develop#egg=aws_proxies
```

If using a requirement file, add the following in the file:

`-e git+git@github.com:davidlonjon/aws-proxies.git@develop#egg=aws_proxies`


## Usage

```
from aws_proxies.proxies import Proxies

# This is the first required step
proxies = Proxies(profile='put_your_aws_profile')
# or if want to see more output use:
# proxies = Proxies(profile='put_your_aws_profile', log_level=20)


# To create the proxies
# Note that the tinyproxy image name need to exists in your owned AMI
# Tinyproxy needs to run on port 8888
# This will create 1 micro instance bound to 4 public IP addresses in the
# 15.0.0.0/16 VPC. The VPC, subnets, internet gateways will all be created automatically.

proxies_config = {
    "available_ips": 4,
    "instances_config": [
        {
            'InstanceType': 't1.micro',
            'ImageName': 'tinyproxy',
            'VPCCidrBlock': '15.0.0.0/16',
            'CidrBlockFormatting': '15.0.\{0\}.\{1\}',
            'SecurityGroups': [
                {
                    'GroupName': 'default',
                    'Description': 'Security group for proxies',
                    'IngressRules': [
                        {
                            'IpProtocol': 'tcp',
                            'FromPort': 8888,
                            'ToPort': 8888,
                            'IpRanges': [
                                {
                                    'CidrIp': '0.0.0.0/0'
                                },
                            ]
                        },
                    ]
                }
            ]
        }
    ]
}

proxies.create(proxies_config=proxies_config, ask_confirm=True, silent=False)
# Without confirmation and completely silent
proxies.create(proxies_config=proxies_config, ask_confirm=False, silent=True)


# To get the public and private ips of the proxies
ips = proxies.instances.get_running_proxies_ips(silent=False)
# ips will be a list of tuples such as [(public_ip, private_ip), (public_ip, private_ip), ...]

# To terminate the proxies and the VPC infrastructure
proxies.delete(ask_confirm=False, silent=True)

```

## Contributing
1. Fork it!
2. Create your feature branch: `git checkout -b my-new-feature`
3. Commit your changes: `git commit -am 'Add some feature'`
4. Push to the branch: `git push origin my-new-feature`
5. Submit a pull request :D

CHANGELOG.md
# Changelog

## 0.1.1 (2016-02-11)

### Features

* Add a readme.rst file. [David Lonjon]

* Add a script to create the rst version of the readme. [David Lonjon]

### Updates

* Update the package version. [David Lonjon]

* Modify the changelog sections name. [David Lonjon]

* Rename the readme file extension to be .md. [David Lonjon]

### Other

* Update changelog and add the rst version. [David Lonjon]

* Add a script to create the changelog which will also create the rst version of it. [David Lonjon]

* Update the requirements files and add gitchangelog as a local requirement. [David Lonjon]

* Add a changelog. [David Lonjon]

* Add .gitchangelog.rc to help generating a changelog. [David Lonjon]


## 0.1.0 (2016-02-11)

### Other

* Rename the aws_proxies module to just proxies. [David Lonjon]

* Cleanup git ignore removing rules to remove bloating. [David Lonjon]

* Removing uneeded files for the package. [David Lonjon]

* Add python package ignore rules. [David Lonjon]

* Add a setup.py for packaging. [David Lonjon]

* Update the README with filler content. [David Lonjon]

* Rename README.md to README.rst. [David Lonjon]

* Add a MANIFEST.in file. [David Lonjon]

* Add a MIT License. [David Lonjon]

* Remove dependencies not needed. [David Lonjon]

* Move symaps_proxies into aws_proxies. [David Lonjon]

* Add a method to get all ip addresses for running proxies. [David Lonjon]

* Adding vpc and instances creation and deletion comfirmation as well printing progress. [David Lonjon]

* Fix code indentation. [David Lonjon]

* Rename variable for consistency and fix comments. [David Lonjon]

* Implement better logging level. [David Lonjon]

* Refactor code, consolidate and simpliy config for proxies creation. [David Lonjon]

* Remove the use of a local settings feature as this should become a package. [David Lonjon]

* Remove commented code. [David Lonjon]

* Make the setup_instances_groups_config method static and add comments. [David Lonjon]

* Create an instances module to simplify code structure and maintenance. [David Lonjon]

* Refactor and introduce a base resources class so other aws reeources can inherit from. [David Lonjon]

* Create a netwok interfaces module to simplify code structure and maintenance. [David Lonjon]

* Move some methods to utils modules and make some methods static methods. [David Lonjon]

* Remove witespaces. [David Lonjon]

* Standardize the way to delete route tables. [David Lonjon]

* Create a netwok acls module to simplify code structure and maintenance. [David Lonjon]

* Rename variable for consistency. [David Lonjon]

* Rename variable for consistency. [David Lonjon]

* Create a route tables module to simplify code structure and maintenance. [David Lonjon]

* Create a security groups module to simplify code structure and maintenance. [David Lonjon]

* Fix comments. [David Lonjon]

* Create a subnets module to simplify code structure and maintenance. [David Lonjon]

* Create an internet gateways module to simplify code structure and maintenance. [David Lonjon]

* Add a missing line return. [David Lonjon]

* Create a vpcs module to simplify code structure and maintenance. [David Lonjon]

* Add comments and rename variable for comprehension. [David Lonjon]

* Move the create_name_tag_for_resource and tag_with_name_with_suffix methods to the utils module. [David Lonjon]

* Add the missing utils module. [David Lonjon]

* Move the filter_resources method to the utils module. [David Lonjon]

* Move the merge_dicts and merge_config methods to the utils module. [David Lonjon]

* Move the query_yes_no method to utils module. [David Lonjon]

* Move create_suffix method to the utils module. [David Lonjon]

* Create a utils module and move the setting up of logger in that module. [David Lonjon]

* Simplify code. [David Lonjon]

* Move proxies class and rename it. [David Lonjon]

* Simplify exception catching. [David Lonjon]

* Simplify AWS interface initialisation, rename constant and cleanup local distribution settings. [David Lonjon]

* Remove unused import. [David Lonjon]

* Remove references to config folder and printing config into file not needed. [David Lonjon]

* Better structure code and add comments. [David Lonjon]

* Add a method to delete internet gateways and re-order the methods to delete vpcs resources to avoid errors. [David Lonjon]

* Add a method to delete internet gateways and re-order the methods to delete vpcs resources to avoid errors. [David Lonjon]

* Add a method to delete subnets. [David Lonjon]

* Add a method to delete security groups. [David Lonjon]

* Fix a typo. [David Lonjon]

* Create methods to delete network acls and route tables. [David Lonjon]

* Add methods to delete elastic network interfaces and terminate instances. [David Lonjon]

* Remove WIP method and optimise code for filtering enis while deleting public ips. [David Lonjon]

* Add a method to release public ips. [David Lonjon]

* Re-arrange code flow for better understanding. [David Lonjon]

* Fix pep8 code formatting. [David Lonjon]

* Automate more the creation of the user data script for running the instance. [David Lonjon]

* Add user data to instance creation. [David Lonjon]

* Add methods to assign public ips to enis and create instances. [David Lonjon]

* Fix a property name of the config object. [David Lonjon]

* Update the port to 8888 for one of the rule in the security group. [David Lonjon]

* Change some error logging to error raising. [David Lonjon]

* Rename varaibles for better understanding. [David Lonjon]

* Remove left over variable assignment used for testing. [David Lonjon]

* Fix PEP8 formatting errors. [David Lonjon]

* Implement better error catching. [David Lonjon]

* Fix info message formatting. [David Lonjon]

* Standardize the usage of single and double quotes. [David Lonjon]

* Add a feature to check image virtualization type against instance types. [David Lonjon]

* Fix the creation of network interfaces. [David Lonjon]


aws_proxies/main.py
# -*- coding: utf-8 -*-

from aws_proxies import AWSProxies
import logging
import sys

proxies_config = {
    "available_ips": 4,
    "instances_config": [
        {
            'InstanceType': 't1.micro',
            'ImageName': 'tinyproxy',
            'VPCCidrBlock': '15.0.0.0/16',
            'CidrBlockFormatting': '15.0.\{0\}.\{1\}',
            'SecurityGroups': [
                {
                    'GroupName': 'default',
                    'Description': 'Security group for proxies',
                    'IngressRules': [
                        {
                            'IpProtocol': 'tcp',
                            'FromPort': 8888,
                            'ToPort': 8888,
                            'IpRanges': [
                                {
                                    'CidrIp': '0.0.0.0/0'
                                },
                            ]
                        },
                        {
                            'IpProtocol': 'tcp',
                            'FromPort': 22,
                            'ToPort': 22,
                            'IpRanges': [
                                {
                                    'CidrIp': '0.0.0.0/0',
                                },
                            ]
                        }
                    ]
                }
            ]
        }
    ]
}


def setup_logger():
        """Setup logger

        Returns:
            object: Logger
        """
        try:  # Python 2.7+
            from logging import NullHandler
        except ImportError:
            class NullHandler(logging.Handler):

                def emit(self, record):
                    pass

        logging.getLogger(__name__).addHandler(NullHandler())
        logging.basicConfig(level=logging.INFO)

        # Raise other modules log levels to make the logs for this module less
        # cluttered with noise
        for _ in ("boto3", "botocore"):
            logging.getLogger(_).setLevel(logging.WARNING)

        return logging.getLogger(__name__)


def main():
    """Main
    """

    logger = setup_logger()

    try:
        proxies = AWSProxies(profile='david_dev', log_level=20)
    except Exception as e:
        logger.error("Error: %s", e.message)
        sys.exit()

    # Create Instances Infrastructure
    proxies.create(proxies_config=proxies_config)

if __name__ == "__main__":
    main()
